### ISSUE TITLE
test_svmoutput_with_type fails in CI builds

### ISSUE THREAD

**Issue by indhub** [apache/incubator-mxnet#8288]
## Description
test_operator_gpu.test_svmoutput_with_type fails in CI master builds. 
https://builds.apache.org/blue/organizations/jenkins/incubator-mxnet/detail/master/528/pipeline/

I've previously seen this failure in this PR build: https://builds.apache.org/blue/organizations/jenkins/incubator-mxnet/detail/PR-7903/48/pipeline/292

After this PR was merged to master, I'm seeing this error in master too. For that reason, I suspect this is caused by PR #7903.

## Environment info (Required)
CI builds. See above links. One failure was in Python 3 GPU Win. Another was in Python 3 MKLML GPU.

## Build info (Required if built from source)
Commit ID used by the failed build on master: 
9f97dac76e43b2ca0acb09a4ff96d416e9edea60

## Error Message:
```
======================================================================
FAIL: test_operator_gpu.test_svmoutput_with_type
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Anaconda3\envs\py3\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\jenkins_slave\workspace\ut-python-gpu\tests\python\gpu\test_operator_gpu.py", line 997, in test_svmoutput_with_type
    check_consistency(sym, ctx_list)
  File "c:\jenkins_slave\workspace\ut-python-gpu\pkg_vc14_gpu\python\mxnet\test_utils.py", line 1338, in check_consistency
    raise e
  File "c:\jenkins_slave\workspace\ut-python-gpu\pkg_vc14_gpu\python\mxnet\test_utils.py", line 1333, in check_consistency
    equal_nan=equal_nan)
  File "c:\jenkins_slave\workspace\ut-python-gpu\pkg_vc14_gpu\python\mxnet\test_utils.py", line 490, in assert_almost_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
Error 5.000000 exceeds tolerance rtol=0.100000, atol=0.100000.  Location of maximum error:(13, 3), a=0.000000, b=1.000000
 a: array([[-0.,  1.,  0., ...,  1.,  1.,  1.],
       [-1.,  1.,  1., ...,  0.,  1.,  0.],
       [ 1., -1.,  1., ...,  1.,  1.,  1.],...
 b: array([[-0.,  1.,  0., ...,  1.,  1.,  1.],
       [-1.,  1.,  1., ...,  0.,  1.,  0.],
       [ 1., -1.,  1., ...,  1.,  1.,  1.],...
-------------------- >> begin captured stdout << ---------------------
Train Err: ctx 2 vs ctx 0 at svmoutput_data

--------------------- >> end captured stdout << ----------------------
```

## Steps to reproduce
Build and run unit tests

**Comment by cjolivier01** [comment]
I looked at the test and the operator code, and this doesn't intersect with that PR's code changes at all.  It's a relatively discrete test, only calling into the SVMOperator (which, in turn, only used mshadow call-outs) and check_consistency() from test_utils.py which I haven't changed.

Ok, I ran this test in a loop in C++ debug mode and it fails about 1% of the time

**Comment by cjolivier01** [comment]
/usr/bin/python3 ./tests/python/gpu/test_operator_gpu.py
0...
5...
10...
15...
20...
25...
30...
35...
40...
Train Err: ctx 2 vs ctx 0 at svmoutput_data
Traceback (most recent call last):
  File "/home/coolivie/src/mxnet/python/mxnet/test_utils.py", line 1333, in check_consistency
    equal_nan=equal_nan)
  File "/home/coolivie/src/mxnet/python/mxnet/test_utils.py", line 490, in assert_almost_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
Error 5.000000 exceeds tolerance rtol=0.100000, atol=0.100000.  Location of maximum error:(3, 3), a=0.000000, b=1.000000
 a: array([[-1.,  1.,  1., ...,  1.,  1.,  1.],
       [-0.,  1.,  0., ...,  0.,  0.,  0.],
       [-1.,  0.,  1., ...,  1.,  1.,  1.],...
 b: array([[-1.,  1.,  1., ...,  1.,  1.,  1.],
       [-0.,  1.,  0., ...,  0.,  0.,  0.],
       [-1.,  0.,  1., ...,  1.,  1.,  1.],...
*** Error in `/usr/bin/python3': munmap_chunk(): invalid pointer: 0x00000000028c84e0 ***

**Comment by cjolivier01** [comment]
/usr/bin/python3 ./tests/python/gpu/test_operator_gpu.py
0...
5...
10...
15...
20...
25...
30...
35...
40...
45...
50...
55...
60...
65...
70...
75...
80...
85...
90...
95...
100...
105...
110...
115...
120...
125...
130...
135...
140...
145...
150...
155...
Train Err: ctx 2 vs ctx 0 at svmoutput_data
Traceback (most recent call last):
  File "/home/coolivie/src/mxnet/python/mxnet/test_utils.py", line 1333, in check_consistency
    equal_nan=equal_nan)
  File "/home/coolivie/src/mxnet/python/mxnet/test_utils.py", line 490, in assert_almost_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
Error 5.000000 exceeds tolerance rtol=0.100000, atol=0.100000.  Location of maximum error:(15, 3), a=0.000000, b=1.000000
 a: array([[ 1.,  1.,  1., ...,  1.,  1.,  1.],
       [-1.,  1.,  1., ...,  1.,  1.,  1.],
       [-0.,  0.,  0., ...,  1.,  1.,  1.],...
 b: array([[ 1.,  1.,  1., ...,  1.,  1.,  1.],
       [-1.,  1.,  1., ...,  1.,  1.,  1.],
       [-0.,  0.,  0., ...,  1.,  1.,  1.],...
*** Error in `/usr/bin/python3': munmap_chunk(): invalid pointer: 0x0000000003db5480 ***

**Comment by cjolivier01** [comment]
Also fails with similar regularity in v0.11.0:

/usr/bin/python3 ./tests/python/gpu/test_operator_gpu.py
0...
5...
10...
15...
20...
25...
30...
35...
40...
45...
50...
55...
60...
65...
70...
75...
80...
85...
90...
Train Err: ctx 2 vs ctx 0 at svmoutput_data
Traceback (most recent call last):
  File "/home/coolivie/src/DeepLearning/mxnet/python/mxnet/test_utils.py", line 912, in check_consistency
    assert_almost_equal(arr, gtarr, rtol=tol[dtypes[i]], atol=tol[dtypes[i]])
  File "/home/coolivie/src/DeepLearning/mxnet/python/mxnet/test_utils.py", line 171, in assert_almost_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
Error 5.000000 exceeds tolerance rtol=0.100000, atol=0.100000.  Location of maximum error:(8, 4), a=0.000000, b=1.000000
 a: array([[-1.,  1.,  1., ...,  0.,  1.,  1.],
       [ 1., -1.,  1., ...,  1.,  0.,  1.],
       [ 0., -1.,  1., ...,  1.,  0.,  1.],...
 b: array([[-1.,  1.,  1., ...,  0.,  1.,  1.],
       [ 1., -1.,  1., ...,  1.,  0.,  1.],
       [ 0., -1.,  1., ...,  1.,  0.,  1.],...
Traceback (most recent call last):
  File "./tests/python/gpu/test_operator_gpu.py", line 1356, in <module>
    test_svmoutput_with_type()
  File "./tests/python/gpu/test_operator_gpu.py", line 973, in test_svmoutput_with_type
    check_consistency(sym, ctx_list)
  File "/home/coolivie/src/DeepLearning/mxnet/python/mxnet/test_utils.py", line 917, in check_consistency
    raise e
  File "/home/coolivie/src/DeepLearning/mxnet/python/mxnet/test_utils.py", line 912, in check_consistency
    assert_almost_equal(arr, gtarr, rtol=tol[dtypes[i]], atol=tol[dtypes[i]])
  File "/home/coolivie/src/DeepLearning/mxnet/python/mxnet/test_utils.py", line 171, in assert_almost_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
Error 5.000000 exceeds tolerance rtol=0.100000, atol=0.100000.  Location of maximum error:(8, 4), a=0.000000, b=1.000000
 a: array([[-1.,  1.,  1., ...,  0.,  1.,  1.],
       [ 1., -1.,  1., ...,  1.,  0.,  1.],
       [ 0., -1.,  1., ...,  1.,  0.,  1.],...
 b: array([[-1.,  1.,  1., ...,  0.,  1.,  1.],
       [ 1., -1.,  1., ...,  1.,  0.,  1.],
       [ 0., -1.,  1., ...,  1.,  0.,  1.],...

**Comment by marcoabreu** [comment]
Still fails on v1.0.0 

```
test_operator_gpu.test_svmoutput_with_type ... Traceback (most recent call last):
  File "/workspace/python/mxnet/test_utils.py", line 1333, in check_consistency
    equal_nan=equal_nan)
  File "/workspace/python/mxnet/test_utils.py", line 490, in assert_almost_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
Error 5.000000 exceeds tolerance rtol=0.100000, atol=0.100000.  Location of maximum error:(5, 2), a=0.000000, b=1.000000
 a: array([[ 1., -0.,  1., ...,  1.,  1.,  1.],
       [-1.,  1.,  1., ...,  1.,  1.,  1.],
       [-1.,  1.,  0., ...,  0.,  0.,  1.],...
 b: array([[ 1., -0.,  1., ...,  1.,  1.,  1.],
       [-1.,  1.,  1., ...,  1.,  1.,  1.],
       [-1.,  1.,  0., ...,  0.,  0.,  1.],...
```

**Comment by vrakesh** [comment]
The random test is failing only when doing float16 tests,
Tried this on varying shapes and different batch runs, without float16 , the test consistently passes even after 25000 runs. Investigating more

**Comment by anirudh2290** [comment]
Reproducible script :

```
import mxnet as mx
import numpy as np

def try_svm(dtype, x):
    exe = sym.simple_bind(grad_req='write',**{'ctx': mx.cpu(0), 'svmoutput_data': (4, 2), 'type_dict': {'svmoutput_data': dtype} })
    exe.arg_arrays[0][:] = x.astype(dtype)
    exe.forward(is_train=True)
    exe.backward(exe.outputs)
    print(exe.arg_arrays[0].asnumpy())
    print(exe.grad_arrays[0].asnumpy())

np.random.seed(1640822401)
x = np.random.normal(size=(4, 2), scale=1.0)
sym = mx.sym.SVMOutput(name='svmoutput', use_linear=True)
try_svm(np.float32, x)
try_svm(np.float16, x)
```

You will find that the two grad_arrays are different.

**Comment by vrakesh** [comment]
The above script from @anirudh2290 , was expanded with a test for gpu as well
```python
import mxnet as mx
import numpy as np


def try_svm(dtype, x):
    exe = sym.simple_bind(grad_req='write',**{'ctx': mx.cpu(0), 'svmoutput_data': (4, 2), 'type_dict': {'svmoutput_data': dtype} })
    exe.arg_arrays[0][:] = x.astype(dtype)
    exe.forward(is_train=True)
    exe.backward(exe.outputs)
    print(exe.outputs[0].asnumpy())
    print(exe.arg_arrays[0].asnumpy())
    print(exe.grad_arrays[0].asnumpy())
def try_svm_gpu(dtype, x):
    exe = sym.simple_bind(grad_req='write',**{'ctx': mx.gpu(0), 'svmoutput_data': (4, 2), 'type_dict': {'svmoutput_data': dtype} })
    exe.arg_arrays[0][:] = x.astype(dtype)
    exe.forward(is_train=True)
    exe.backward(exe.outputs)
    print(exe.outputs[0].asnumpy())
    print(exe.arg_arrays[0].asnumpy())
    print(exe.grad_arrays[0].asnumpy())

print("CPU run\n")
np.random.seed(1640822401)
x = np.random.normal(size=(4, 2), scale=1.0)
sym = mx.sym.SVMOutput(name='svmoutput', use_linear=True)
try_svm(np.float32, x)
try_svm(np.float16, x)

print("GPU run\n")
np.random.seed(643549208)
try_svm_gpu(np.float32, x)
try_svm_gpu(np.float16, x)
```
The output of the test is

```bash
CPU run
# outputs, args float 32
[[ 0.23923199 -0.5441972 ]
 [ 0.30816466 -0.34832394]
 [ 0.99995518  1.06915462]
 [-0.37221318 -0.06465939]]
[[ 0.23923199 -0.5441972 ]
 [ 0.30816466 -0.34832394]
 [ 0.99995518  1.06915462]
 [-0.37221318 -0.06465939]]

# Grads Float32
[[-1.  1.]
 [-1.  1.]
 [-1.  1.]
 [-1.  1.]]

# outputs, args float16
[[ 0.23925781 -0.54443359]
 [ 0.30810547 -0.34838867]
 [ 1.          1.06933594]
 [-0.37231445 -0.06463623]]
[[ 0.23925781 -0.54443359]
 [ 0.30810547 -0.34838867]
 [ 1.          1.06933594]
 [-0.37231445 -0.06463623]]

#Grads float 16
[[-1.  1.]
 [-1.  1.]
 [-0.  1.]
 [-1.  1.]]
GPU run

# outputs, args 
[[ 0.23923199 -0.5441972 ]
 [ 0.30816466 -0.34832394]
 [ 0.99995518  1.06915462]
 [-0.37221318 -0.06465939]]
[[ 0.23923199 -0.5441972 ]
 [ 0.30816466 -0.34832394]
 [ 0.99995518  1.06915462]
 [-0.37221318 -0.06465939]]

# Grad float 32
[[-1.  1.]
 [-1.  1.]
 [-1.  1.]
 [-1.  1.]]

#outputs float 16
[[ 0.23925781 -0.54443359]
 [ 0.30810547 -0.34838867]
 [ 1.          1.06933594]
 [-0.37231445 -0.06463623]]
[[ 0.23925781 -0.54443359]
 [ 0.30810547 -0.34838867]
 [ 1.          1.06933594]
 [-0.37231445 -0.06463623]]

#grad float 16
[[-1.  1.]
 [-1.  1.]
 [-0.  1.]
 [-1.  1.]]
```

As seen above for float 16, the output that is close to 0.999 rounds to 1.0. Our default margin for SVMOutput layer is 1.0

In L1_SVM function of the operator we compare margin > src (where src is the outputs) 

For float32 this resolves to 1.0 > 0.99532
For float16 this resolves to 1.0 > 1.0

Thus showing wrong results in grad (Type casted from bool to float) which is 1.0 and 0.0 thus resulting in the error

To fix this the input being provided for test, should not be normally distributed, as values closer to the margin, will break the condition due to rounding at float16 level.

One solution is to provide the test with a uniform distribution of inputs , that will never get close to the margin.
Creating a PR for this

**Comment by marcoabreu** [comment]
Closed as of https://github.com/apache/incubator-mxnet/pull/11505
